workflow-trace-viewer/src/jvmMain/kotlin/com/squareup/workflow1/traceviewer/util/SocketClient.kt

```diff
@@ -0,0 +1,69 @@
+package com.squareup.workflow1.traceviewer.util
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
+import java.net.Socket
+import java.net.SocketException
+
+/**
+ * This is a client that connects to the `ActionLogger` Unix Domain Socket and listens for any new
+ * render passes. Since this app is on JVM and the server is on Android, we use ADB to forward the
+ * port onto the socket.
+ */
+internal class SocketClient {
+  private lateinit var socket: Socket
+  private var initialized = false
+  val renderPassChannel: Channel<String> = Channel(Channel.UNLIMITED)
+
+  /**
+   * We use any available ports on the host machine to connect to the emulator.
+   *
+   * `workflow-trace` is the name of the unix socket created, and since Android uses
+   * `LocalServerSocket` -- which creates a unix socket on the linux abstract namespace -- we use
+   * `localabstract:` to connect to it.
+   */
+  fun start() {
+    initialized = true
+    val process = ProcessBuilder(
+      "adb", "forward", "tcp:0", "localabstract:workflow-trace"
+    ).start()
+
+    // The adb forward command will output the port number it picks to connect.
+    val port = run {
+      process.waitFor()
+      process.inputStream.bufferedReader().readText()
+        .trim().toInt()
+    }
```

Comment:
nit: thoughts making a function call out instead of this run syntax.

readInt() or something like that 

where readInt == 

```
fun readInt() {
      process.waitFor()
      process.inputStream.bufferedReader().readText()
        .trim().toInt()
    }
```